File: app/src/main/java/ai/kortnevdmitriy/msafiri/activities/Signin.java
package ai.kortnevdmitriy.msafiri.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.SignInButton;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.tapadoo.alerter.Alerter;

import ai.kortnevdmitriy.msafiri.R;

public class Signin extends AppCompatActivity implements GoogleApiClient.OnConnectionFailedListener, View.OnClickListener {
    private static final String TAG = "SignInActivity";
    private static final int RC_SIGN_IN = 9001;
    private GoogleApiClient mGoogleApiClient;
    private SignInButton mSignInButton;
    private FirebaseAuth auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        auth = FirebaseAuth.getInstance(); // Get Firebase auth instance
        if (auth.getCurrentUser() != null) {
            startActivity(new Intent(getApplicationContext(), Home.class));
            finish();
        }

        // Make Screen Window Fullscreen before setting the contents.
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);

        setContentView(R.layout.activity_signin);

        mSignInButton = findViewById(R.id.sign_in_button);
        mSignInButton.setSize(SignInButton.SIZE_WIDE);
        mSignInButton.setOnClickListener(this);
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();
        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this /* FragmentActivity */, this /* OnConnectionFailedListener */)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();
        auth = FirebaseAuth.getInstance();
    }


    private void handleFirebaseAuthResult(AuthResult authResult) {
        if (authResult != null) {
            // Welcome the user
            FirebaseUser user = authResult.getUser();
            Toast.makeText(this, "Welcome " + user.getDisplayName(), Toast.LENGTH_SHORT).show();

            // Go back to the main activity
            startActivity(new Intent(getApplicationContext(), Home.class));
        }

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.sign_in_button:
                signIn();
                break;
            default:
                return;
        }
    }

    private void signIn() {
        Intent signInIntent = Auth.GoogleSignInApi.getSignInIntent(mGoogleApiClient);
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Result returned from launching the Intent from GoogleSignInApi.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            GoogleSignInResult result = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
            if (result.isSuccess()) {
                // Google Sign In was successful, authenticate with Firebase
                GoogleSignInAccount account = result.getSignInAccount();
                firebaseAuthWithGoogle(account);
            } else {
                // Google Sign In failed
                googleSignInErrorNotification();
                Log.e(TAG, "Google Sign In failed.");
            }
        }
    }

    private void firebaseAuthWithGoogle(GoogleSignInAccount acct) {
        Log.d(TAG, "firebaseAuthWithGoogle:" + acct.getId());
        AuthCredential credential = GoogleAuthProvider.getCredential(acct.getIdToken(), null);
        auth.signInWithCredential(credential)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        Log.d(TAG, "signInWithCredential:onComplete:" + task.isSuccessful());

                        // If sign in fails, display a message to the user. If sign in succeeds
                        // the auth state listener will be notified and logic to handle the
                        // signed in user can be handled in the listener.
                        if (!task.isSuccessful()) {
                            Log.w(TAG, "signInWithCredential", task.getException());
                            authenticationErrorNotification();
                        } else {
                            startActivity(new Intent(getApplicationContext(), Home.class));
                            finish();
                        }
                    }
                });
    }


    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
        // An unresolvable error has occurred and Google APIs (including Sign-In) will not
        // be available.
        Log.d(TAG, "onConnectionFailed:" + connectionResult);
        Alerter.create(this)
                .setTitle("Google Play")
                .setText("Google Play Services error")
                .setBackgroundColorRes(android.R.color.holo_red_dark) // or setBackgroundColorInt(Color.CYAN)
                .show();
    }

    public void authenticationErrorNotification() {
        Alerter.create(this)
                .setTitle("Authentication")
                .setText("Authentication failed. Check your Internet")
                .setBackgroundColorRes(android.R.color.holo_red_dark) // or setBackgroundColorInt(Color.CYAN)
                .show();
    }

    public void googleSignInErrorNotification() {
        Alerter.create(this)
                .setTitle("Google Signin")
                .setText("Google Sign In failed")
                .setBackgroundColorRes(android.R.color.holo_red_dark) // or setBackgroundColorInt(Color.CYAN)
                .show();
    }
}

File: app/src/main/java/ai/kortnevdmitriy/msafiri/activities/Signin.kt
package ai.kortnevdmitriy.msafiri.activities

import ai.kortnevdmitriy.msafiri.R
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.SignInButton
import com.google.android.gms.common.api.GoogleApiClient
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.tapadoo.alerter.Alerter

class Signin : AppCompatActivity(), GoogleApiClient.OnConnectionFailedListener,
	View.OnClickListener {
	private var mGoogleApiClient: GoogleApiClient? = null
	private var mSignInButton: SignInButton? = null
	private var auth: FirebaseAuth? = null
	
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		auth = FirebaseAuth.getInstance() // Get Firebase auth instance
		if (auth?.currentUser != null) {
			startActivity(Intent(applicationContext, Home::class.java))
			finish()
		}
		
		// Make Screen Window Fullscreen before setting the contents.
		requestWindowFeature(Window.FEATURE_NO_TITLE)
		window.setFlags(
			WindowManager.LayoutParams.FLAG_FULLSCREEN,
			WindowManager.LayoutParams.FLAG_FULLSCREEN
		)
		
		setContentView(R.layout.activity_signin)
		
		mSignInButton = findViewById(R.id.sign_in_button)
		mSignInButton?.setSize(SignInButton.SIZE_WIDE)
		mSignInButton?.setOnClickListener(this)
		val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
			//.requestIdToken(getString(R.string.default_web_client_id))
			.requestEmail()
			.build()
		mGoogleApiClient = GoogleApiClient.Builder(this)
			.enableAutoManage(this /* FragmentActivity */, this /* OnConnectionFailedListener */)
			.addApi(Auth.GOOGLE_SIGN_IN_API, gso)
			.build()
		auth = FirebaseAuth.getInstance()
	}
	
	
	private fun handleFirebaseAuthResult(authResult: AuthResult?) {
		if (authResult != null) {
			// Welcome the user
			val user = authResult.user
			Toast.makeText(this, "Welcome " + user.displayName!!, Toast.LENGTH_SHORT).show()
			
			// Go back to the main activity
			startActivity(Intent(applicationContext, Home::class.java))
		}
		
	}
	
	override fun onClick(view: View) {
		when (view.id) {
			R.id.sign_in_button -> signIn()
			else -> return
		}
	}
	
	private fun signIn() {
		val signInIntent = Auth.GoogleSignInApi.getSignInIntent(mGoogleApiClient)
		startActivityForResult(signInIntent, RC_SIGN_IN)
	}
	
	public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
		super.onActivityResult(requestCode, resultCode, data)
		
		// Result returned from launching the Intent from GoogleSignInApi.getSignInIntent(...);
		if (requestCode == RC_SIGN_IN) {
			val result = Auth.GoogleSignInApi.getSignInResultFromIntent(data)
			if (result.isSuccess) {
				// Google Sign In was successful, authenticate with Firebase
				val account = result.signInAccount
				firebaseAuthWithGoogle(account!!)
			} else {
				// Google Sign In failed
				googleSignInErrorNotification()
				Log.e(TAG, "Google Sign In failed.")
			}
		}
	}
	
	private fun firebaseAuthWithGoogle(acct: GoogleSignInAccount) {
		Log.d(TAG, "firebaseAuthWithGoogle:" + acct.id!!)
		val credential = GoogleAuthProvider.getCredential(acct.idToken, null)
		auth!!.signInWithCredential(credential)
			.addOnCompleteListener(this) { task ->
				Log.d(TAG, "signInWithCredential:onComplete:" + task.isSuccessful)
				
				// If sign in fails, display a message to the user. If sign in succeeds
				// the auth state listener will be notified and logic to handle the
				// signed in user can be handled in the listener.
				if (!task.isSuccessful) {
					Log.w(TAG, "signInWithCredential", task.exception)
					authenticationErrorNotification()
				} else {
					startActivity(Intent(applicationContext, Home::class.java))
					finish()
				}
			}
	}
	
	
	override fun onConnectionFailed(connectionResult: ConnectionResult) {
		// An unresolvable error has occurred and Google APIs (including Sign-In) will not
		// be available.
		Log.d(TAG, "onConnectionFailed:$connectionResult")
		Alerter.create(this)
			.setTitle("Google Play")
			.setText("Google Play Services error")
			.setBackgroundColorRes(android.R.color.holo_red_dark) // or setBackgroundColorInt(Color.CYAN)
			.show()
	}
	
	private fun authenticationErrorNotification() {
		Alerter.create(this)
			.setTitle("Authentication")
			.setText("Authentication failed. Check your Internet")
			.setBackgroundColorRes(android.R.color.holo_red_dark) // or setBackgroundColorInt(Color.CYAN)
			.show()
	}
	
	private fun googleSignInErrorNotification() {
		Alerter.create(this)
			.setTitle("Google Signin")
			.setText("Google Sign In failed")
			.setBackgroundColorRes(android.R.color.holo_red_dark) // or setBackgroundColorInt(Color.CYAN)
			.show()
	}
	
	companion object {
		private const val TAG = "SignInActivity"
		private const val RC_SIGN_IN = 9001
	}
}
